### PR TITLE
Add workflow-role based authorization

### PR DESCRIPTION
**Pull request by straleyb**
Add the ability to allow a user to see or not see a work based on their workflow roles and the current workflow state.

### CODE REVIEW

**Comment by jcoyne** [comment]
I don't understand the purpose of adding entities to these feature tests. This just slows down the test suite.

**Comment by straleyb** [comment]
@jcoyne Sipity based access control relies on a Sipity::Entity existing in the database for a workflow. This is the line of code here https://github.com/projecthydra/curation_concerns/pull/1103/files#diff-aede147ec36a958b2d6485797b8b15c6R19

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+module CurationConcerns
+  module FilterSuppressedWithRoles
+    extend CurationConcerns::FilterSuppressed
```
Comment (jcoyne) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: I don't think this is doing what you think it's doing

File: app/models/concerns/curation_concerns/suppressible.rb [hunk 2]
```diff
@@ -1,20 +1,25 @@
...
+    end
+
+    def deactivate
+      self.state = Vocab::FedoraResourceStatus.inactive
+    end
 
-      class_attribute :state_workflow, instance_writer: false
-      self.state_workflow = StateWorkflow
+    def activate
+      self.state = Vocab::FedoraResourceStatus.active
```
Comment (jcoyne) [inline app/models/concerns/curation_concerns/suppressible.rb]: I think this should remain in the ActivateObject service class. Why are you moving this here. The responsibity of the model is persisting data. The responsibility of the service object is knowing how to activate. Each object should have a single responsibility.

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb [hunk 3]
```diff
@@ -4,7 +4,16 @@
 # which is included into .internal_test_app/app/controllers/generic_works_controller.rb
 describe CurationConcerns::GenericWorksController do
   let(:user) { create(:user) }
-  before { sign_in user }
+  let(:work) { create(:generic_work, user: user) }
+  let(:sipity_entity) do
```
Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: If you change this `let` to `let!`, you can remove `sipity_entity` on line 15 below.

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb [hunk 4]
```diff
@@ -4,7 +4,16 @@
...
+    create(:sipity_entity,
+           proxy_for_global_id: work.to_global_id.to_s,
+           workflow_state_id: 2)
```
Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: As discussed in Slack, you should be able to :fire: all the hard-coded `workflow_state_id` stuff you added.

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb [hunk 5]
```diff
@@ -4,7 +4,16 @@
...
+  end
+  before do
+    sign_in user
```
Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: Why would this be required now, if it wasn't required before?

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb [hunk 6]
```diff
@@ -16,7 +25,13 @@
 
       context "with a parent work" do
         let(:parent) { create(:generic_work, title: ['Parent Work'], user: user, ordered_members: [work]) }
+        let(:parent_sipity_entity) do
```
Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: If you use `let!` instead of `let`, you can dump the change to line 34 below.

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb [hunk 7]
```diff
@@ -238,55 +252,56 @@
   end
 
   describe '#destroy' do
-    let(:work_to_be_deleted) { create(:private_generic_work, user: user) }
```
Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: Why rename the variable here? Do you want this `work` to stand in for the `work` created on line 7?

File: spec/factories/generic_works.rb [hunk 8]
```diff
@@ -4,6 +4,7 @@
   factory :generic_work, aliases: [:work, :private_generic_work], class: GenericWork do
     transient do
       user { FactoryGirl.create(:user) }
+      # workflow_state { FactoryGirl.create(:workflow_state) }
```
Comment (mjgiarlo) [inline spec/factories/generic_works.rb]: Remove if not needed when you're ready to submit the PR.

File: spec/factories/generic_works.rb [hunk 9]
```diff
@@ -13,6 +14,10 @@
       work.apply_depositor_metadata(evaluator.user.user_key)
     end
 
+    # after(:create) do |work, evaluator|
```
Comment (mjgiarlo) [inline spec/factories/generic_works.rb]: Remove these lines if not needed when you're ready to submit the PR.

File: spec/factories/workflows.rb [hunk 10]
```diff
@@ -1,5 +1,6 @@
 FactoryGirl.define do
   factory :workflow, class: Sipity::Workflow do
-    name 'generic_work'
+    sequence(:name) { |n| "generic_work-#{n}" }
+    # name 'generic_work'
```
Comment (mjgiarlo) [inline spec/factories/workflows.rb]: Remove if not needed when you're ready to submit the PR.

File: spec/factories/workflows.rb [hunk 11]
```diff
@@ -1,5 +1,6 @@
 FactoryGirl.define do
   factory :workflow, class: Sipity::Workflow do
-    name 'generic_work'
+    sequence(:name) { |n| "generic_work-#{n}" }
```
Comment (mjgiarlo) [inline spec/factories/workflows.rb]: Is this change still needed?

File: spec/features/add_file_spec.rb [hunk 12]
```diff
@@ -3,9 +3,15 @@
 feature 'Add an attached file' do
   let(:user) { create(:user) }
   let!(:work) { create(:work, user: user) }
+  let(:sipity_entity) do
```
Comment (mjgiarlo) [inline spec/features/add_file_spec.rb]: Use `let!` instead of `let` and you can remove line 14 below.

File: spec/features/add_file_spec.rb [hunk 13]
```diff
@@ -3,9 +3,15 @@
...
+    create(:sipity_entity,
+           proxy_for_global_id: work.to_global_id.to_s,
+           workflow_state_id: 2)
```
Comment (mjgiarlo) [inline spec/features/add_file_spec.rb]: :fire:

File: spec/features/create_child_work_spec.rb [hunk 14]
```diff
@@ -3,7 +3,11 @@
 
 feature 'Creating a new child Work', :workflow do
   let(:user) { FactoryGirl.create(:user) }
-
+  let(:sipity_entity) do
+    create(:sipity_entity,
```
Comment (mjgiarlo) [inline spec/features/create_child_work_spec.rb]: Same re: above: `let!`

File: spec/features/create_child_work_spec.rb [hunk 15]
```diff
@@ -3,7 +3,11 @@
...
+           proxy_for_global_id: parent.to_global_id.to_s,
+           workflow_state_id: 2)
```
Comment (mjgiarlo) [inline spec/features/create_child_work_spec.rb]: :fire:

File: spec/models/generic_work_spec.rb [hunk 16]
```diff
@@ -37,17 +37,36 @@
...
       it { is_expected.to be false }
     end
   end
 
+  context "when activating or deactivating a work" do
+    let(:work) { described_class.new }
+    describe "#activate" do
+      subject { work.activate }
+      it "sets its state to active" do
+        subject
```
Comment (mjgiarlo) [inline spec/models/generic_work_spec.rb]: Not sure what line 56 is adding. Can you rework this test to be more succinct? Like

```ruby
describe '#activate' do
  subject { work.activate }
  it { is_expected.not_to be_suppressed }
end
```

File: spec/models/generic_work_spec.rb [hunk 17]
```diff
@@ -37,17 +37,36 @@
...
+        expect(work.suppressed?).to be false
+      end
+    end
+
+    describe "#deactivate" do
+      subject { work.deactivate }
+      it "sets its state to inactive" do
+        subject
```
Comment (mjgiarlo) [inline spec/models/generic_work_spec.rb]: Nuke line 64 and rework as above:

```ruby
describe '#deactivate' do
  subject { work.deactivate }
  it { is_expected.to be_suppressed }
end
```

File: spec/services/curation_concerns/workflow/permission_query_spec.rb [hunk 18]
```diff
@@ -24,7 +24,7 @@ module Workflow
                                workflow_state: PowerConverter.convert_to_sipity_workflow_state('initial', scope: sipity_workflow)
                               )
       end
-      let(:sipity_workflow) { Sipity::Workflow.first }
+      let(:sipity_workflow) { Sipity::Workflow.where(name: 'testing').first }
```
Comment (mjgiarlo) [inline spec/services/curation_concerns/workflow/permission_query_spec.rb]: Is this still needed? Are there > 1 workflows?

File: spec/services/curation_concerns/workflow/status_list_service_spec.rb [hunk 19]
```diff
@@ -4,6 +4,7 @@
   describe "#each" do
     let(:user) { create(:user) }
     let(:service) { described_class.new(user) }
+    let(:sipity_entity) { instance_double(Sipity::Entity, workflow: workflow, workflow_state: workflow_state) }
```
Comment (mjgiarlo) [inline spec/services/curation_concerns/workflow/status_list_service_spec.rb]: Why is this test using instance doubles when the others are using factories? If you can make this test use the same factories, that would bring extra consistency.

File: spec/services/curation_concerns/workflow/status_list_service_spec.rb [hunk 20]
```diff
@@ -14,16 +15,21 @@
...
     before do
-      create(:sipity_entity)
       ActiveFedora::SolrService.add([document, ability], commit: true)
     end
 
     let(:results) { service.each.to_a }
 
     context "when user has roles" do
       before do
+        sipity_entity
```
Comment (mjgiarlo) [inline spec/services/curation_concerns/workflow/status_list_service_spec.rb]: Use `let!` above instead.

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb (hunk 3 above)
Comment (jcoyne) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: It's better to use a before block though. Otherwise rubocop rspec will complain.

Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: @jcoyne is that so? There's lots of `let!` in the current codebase.

Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: @jcoyne :ok_hand:

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb (hunk 1 above)
Comment (straleyb) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: So my idea of this was, creating an object that inherits the methods from `CurationConcerns::FilterSuppressed` and overrides them to cut down on code duplication. Would you just include the module instead of extend the module?

**Comment by straleyb** [comment]
@jcoyne Based on your comment about creating objects slowing down the test suite, I hadn't noticed a slow down personally, but would you much rather mocking the entities instead? I had the assumption they needed to exist in the database for the query here. https://github.com/projecthydra/curation_concerns/pull/1103/files#diff-aede147ec36a958b2d6485797b8b15c6R20

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb (hunk 7 above)
Comment (straleyb) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: So I changed the name of this variable for consistency sake. Thats just that many more places you can miss a variable and confused me several times. So I figured it would be nice to go with a more consistent name. But I can change it back no biggy. Was just a proposition really.

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+    def only_active_works(solr_parameters)
+      if current_work.respond_to?(:to_sipity_entity) # Used because Collections use WorkSearchBuilder to filter down to a single item
+        super if user_work_permissions.empty?
```
Comment (mjgiarlo) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Per @jcoyne's above comment, the `super`s in this module do not wind up calling the method with the same name in `FilterSuppressed` unless the class you're calling this method on *also* includes `FilterSuppressed`.

File: spec/factories/workflows.rb (hunk 11 above)
Comment (straleyb) [inline spec/factories/workflows.rb]: Yes, this change is needed. This is because now that we are creating sipity entities, these workflows need to exist and not having this will throw the `UNIQUE constraint failed` errors in the DB

**Comment by jcoyne** [comment]
@straleyb Okay, I wasn't sure why that was required. I think you could test out just the SearchBuilder itself with real objects, and then mock out the response from the search builder on the Controller tests.  Otherwise add comments to the test about why those are needed.

File: spec/controllers/curation_concerns/generic_works_controller_spec.rb (hunk 7 above)
Comment (mjgiarlo) [inline spec/controllers/curation_concerns/generic_works_controller_spec.rb]: I'm OK with this change.

**Comment by straleyb** [comment]
@jcoyne Sounds good. Thanks for the clarification. I can go through and rework those.

File: spec/factories/workflows.rb (hunk 11 above)
Comment (mjgiarlo) [inline spec/factories/workflows.rb]: :ok_hand: sounds good

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb (hunk 21 above)
Comment (straleyb) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: So would the change here be to `include` FilterSuppressed rather than extend it? Or is there another pattern or technique I should be using here?

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb (hunk 1 above)
Comment (mjgiarlo) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: @straleyb I believe that should do the trick.

**Comment by mjgiarlo** [comment]
@jcoyne when you get a minute, could you re-review this PR? It's :green_heart:

File: app/models/concerns/curation_concerns/suppressible.rb [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+    ##
+    # Used to restrict visibility on search results for a work that is inactive. If the state is not set, the
+    # default behavior is to consider the work not to be suppressed.
+    #
+    # Override this method if you have some criteria by which records should not display in the search results.
+    def suppressed?
+      return false if state.nil?
+      # This is a clumsy check only needed under RDF < 2.0
+      if ENV['RDF_VERSION'] < '2.0'
```
Comment (jcoyne) [inline app/models/concerns/curation_concerns/suppressible.rb]: This will only be set on Travis. It won't be suitable in production

Comment (mjgiarlo) [inline app/models/concerns/curation_concerns/suppressible.rb]: Dang. Shall I ask RDF.rb its version instead?

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+      else
+        super
+      end
+    end
+
+    private
+
+      def current_work
+        ActiveFedora::Base.find(blacklight_params[:id])
```
Comment (jcoyne) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Loading from Fedora is a slow operation, which SearchBuilders were created to avoid.  SearchBuilders should never interface with Fedora.

File: app/models/concerns/curation_concerns/suppressible.rb [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+    ##
+    # Used to restrict visibility on search results for a work that is inactive. If the state is not set, the
+    # default behavior is to consider the work not to be suppressed.
+    #
+    # Override this method if you have some criteria by which records should not display in the search results.
+    def suppressed?
+      return false if state.nil?
+      # This is a clumsy check only needed under RDF < 2.0 where URIs
+      # are returned as AT::Resources
```
Comment (jcoyne) [inline app/models/concerns/curation_concerns/suppressible.rb]: I believe both versions return `AT::Resources`, but the latter versions implement `==` correctly

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb [hunk 25]
```diff
@@ -0,0 +1,28 @@
+module CurationConcerns
+  module FilterSuppressedWithRoles
```
Comment (jcoyne) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Add a module comment about when this can be used and its assumptions (`solr_params[:id]` being one).

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+    extend ActiveSupport::Concern
+    include CurationConcerns::FilterSuppressed
+
+    def only_active_works(solr_parameters)
+      if current_work.respond_to?(:to_sipity_entity) # Used because Collections use WorkSearchBuilder to filter down to a single item
```
Comment (jcoyne) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Maybe the Collections shouldn't use WorkSearchBuilder. That's probably better than adding these conditionals.

File: app/services/curation_concerns/workflow/activate_object.rb [hunk 27]
```diff
@@ -1,5 +1,9 @@
 module CurationConcerns
   module Workflow
+    # This is a built in function for workflow, so that a workflow action can be created that
+    # activates an object.
+    # Target is an instance of a model with CurationConcerns::Suppressible
```
Comment (jcoyne) [inline app/services/curation_concerns/workflow/activate_object.rb]: This should be YARD doc on the `.call` method.

File: app/services/curation_concerns/workflow/deactivate_object.rb [hunk 28]
```diff
@@ -2,6 +2,8 @@ module CurationConcerns
   module Workflow
     # This is a built in function for workflow, so that a workflow action can be created that
     # deactivates an object.
+    # Target is an instance of a model with CurationConcerns::Suppressible
```
Comment (jcoyne) [inline app/services/curation_concerns/workflow/deactivate_object.rb]: This should be YARD doc on the `.call` method.

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb (hunk 26 above)
Comment (jcoyne) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Actually, I'm not seeing where this is happening.  Current_work should always be an instance of a work. Can we remove this conditional?

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb (hunk 25 above)
Comment (mjgiarlo) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Done.

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb (hunk 26 above)
Comment (mjgiarlo) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Removed.

File: app/services/curation_concerns/workflow/activate_object.rb (hunk 27 above)
Comment (mjgiarlo) [inline app/services/curation_concerns/workflow/activate_object.rb]: Done.

File: app/services/curation_concerns/workflow/deactivate_object.rb (hunk 28 above)
Comment (mjgiarlo) [inline app/services/curation_concerns/workflow/deactivate_object.rb]: Done.

File: app/search_builders/curation_concerns/filter_suppressed_with_roles.rb [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+  # actions for the work's current state
+  #
+  # Assumes presence of `blacklight_params[:id]` and a SolrDocument
+  # corresponding to that `:id` value
+  module FilterSuppressedWithRoles
+    extend ActiveSupport::Concern
+    include CurationConcerns::FilterSuppressed
+
+    def only_active_works(solr_parameters)
+      super if user_work_permissions.empty?
```
Comment (jcoyne) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Can we put this in a method called `current_user_has_active_role?` and then do: `return super unless current_user_has_active_role?`

Comment (mjgiarlo) [inline app/search_builders/curation_concerns/filter_suppressed_with_roles.rb]: Will do.
